pkg/clustermesh/mcsapi/service_controller.go

```diff
@@ -78,6 +79,26 @@ func servicePorts(svcImport *mcsapiv1alpha1.ServiceImport) []corev1.ServicePort
 			AppProtocol: port.AppProtocol,
 			Port:        port.Port,
 		})
+		portMap[port.Port] = &ports[len(ports)-1]
```

Comment:
Be careful that this will not work if the slice gets resized, because the elements would be copied to a different location, and the pointers would then point to dangling copies. See https://go.dev/play/p/zwLEtSo_0BS for instance.

I'd suggest reversing the logic, constructing the map from the local service ports (if not nil), and then assigning here the `TargetPort` value retrieved from that map (the default value works fine when not present).

Reply:
Oh yes indeed good catch! The slice should never be resized in that function though because it's preallocated but I definitely agree that this is walking on thin ice and it may hit us on any future code change!